File: Single/Medium/The.Gift.java
// The Gift
import java.util.*;

class Solution {
    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        int N = in.nextInt();                   // Number of participants
        int C = in.nextInt();                   // Price of gift

        StringBuffer output = new StringBuffer();
        int sum = 0;
        int budget[] = new int[N];
        for (int i = 0; i < N; i++) {
            int B = in.nextInt();               // Budget
            budget[i] = B;
            sum += B;
        }
        System.err.println(N + " participants, price of gift:" + C + ", sum:" + sum);        
        if (C > sum) {
            output.append("IMPOSSIBLE");
        } else {
            Arrays.sort(budget);
            for (int i = 0; i < budget.length; ++i) {
                int avg = (int)(C / (budget.length - i));
                if (budget[i] < avg) {
                    C -= budget[i];
                    output.append(budget[i]);
                } else {
                    C -= avg;
                    output.append(avg);
                }
                output.append("\n");
            }
        }
        System.err.println("output:");
        System.out.println(output.toString());
    }// main()
}

File: Single/Medium/The.Gift.php
<?php
// The Gift
    fscanf(STDIN, "%d", $N);        // Number of participants
    fscanf(STDIN, "%d", $C);        // Price of gift
    
    $output = "";
    $sum = 0;
    $budget = array();
    for ($i = 0; $i < $N; $i++) {
        fscanf(STDIN, "%d", $B);    // Budget
        $budget[$i] = $B;
        $sum += $B;
    }
    error_log("$N participants, price of gift:$C, sum:$sum");        
    if ($C > $sum) {
        $output = ("IMPOSSIBLE\n");
    } else {
        sort($budget);
        for ($i = 0; $i < count($budget); $i++) {
            $avg = (integer)($C / (count($budget) - $i));
            if ($budget[$i] < $avg) {
                $C -= $budget[$i];
                $output .= ($budget[$i]);
            } else {
                $C -= $avg;
                $output .= ($avg);
            }
            $output .= ("\n");
        }
    }
    error_log("output:\n");
    print($output);    
?>